### DISCUSSION TITLE
Bolt-on spatial feature encoder improves YOLOv8-OBB classification on DOTA without modifying the model

### DISCUSSION THREAD

**Discussion by jackkowalik** [ultralytics/ultralytics#23821] [Show and tell]
## Bolt-on spatial feature encoder improves YOLOv8-OBB classification on DOTA without modifying the model

We built a frozen spatial feature encoder that operates downstream of YOLOv8 detection. It takes each detection crop, produces a 920-dimensional feature vector, and when concatenated with YOLO's softmax output and fed into a lightweight LightGBM classifier, improves classification accuracy. The detection pipeline is completely untouched.

### Results on DOTA v1.0

YOLOv8l-OBB, 50,348 matched detections from 3,143 tiled validation images (458 original scenes), 5-fold scene-level cross-validation:

```
                          YOLO Direct    Bolt-On
Weighted F1               0.9925         0.9929
Accuracy                  0.9925         0.9929
Macro F1                  0.9826         0.9827

Per-class gains:
  helicopter              0.502  →  0.916   (+0.414)
  plane                   0.976  →  0.998   (+0.022)
  basketball-court        0.931  →  0.947   (+0.015)
  soccer-ball-field       0.960  →  0.972   (+0.012)
  tennis-court            0.985  →  0.990   (+0.005)

No class degraded across all 15 categories.
```

The helicopter result is clearly the most impressive. YOLO's softmax alone gives essentially random classification on rotary-wing aircraft (0.502 F1). The encoder captures the rotational symmetry signature of rotor aircraft vs bilateral symmetry of fixed-wing, pushing helicopter F1 to 0.916.

### How it works

1. Run YOLOv8-OBB detection on DOTA tiles normally
2. For each matched detection (IoU ≥ 0.5 with ground truth), crop the region
3. Resize crop to 128×128 grayscale
4. Send crop to the encoder API → get 920-dim feature vector
5. Concatenate YOLO's 15-class softmax with the 920-dim vector
6. Train LightGBM with balanced class weights
7. Evaluate under 5-fold scene-level cross-validation (GroupKFold by original scene)

The encoder is frozen and has never been trained on aerial imagery, satellite data, or any of the DOTA object categories.

**Requirements:**
```
pip install ultralytics lightgbm scikit-learn numpy opencv-python requests pillow tqdm
```

**Data:** DOTA v1.0 validation set (images + labels) from https://captain-whu.github.io/DOTA/

### Other benchmarks with the same frozen encoder

We've validated the same bolt-on pattern across multiple domains and modalities:

```
Benchmark       Domain              Modality          Baseline    Bolt-On     Error Reduction
xView3          Maritime vessels    SAR (C-band)      0.875 F1    0.881 F1    4.6%
DOTA            Aerial objects      HR aerial         0.992 F1    0.993 F1    8.9%
EuroSAT         Land cover          Multispectral     0.983 Acc   0.985 Acc   10.6%
SpaceNet 6      Building extract.   SAR (X-band)      0.835 F1    0.858 F1    14.1%
RarePlanes      Aircraft classif.   VHR satellite     0.660 F1    0.794 F1    39.5%
xView2          Building damage     RGB optical       0.710 F1    0.828 F1    40.7%
```

All six use the identical frozen encoder with no modification between domains.

### Links

- API docs & interface: https://authorize.earth/r&d/spatial
- Contact: jackk@authorize.earth

Looking for people to test this against other models, datasets, and domains. Happy to provision evaluation keys for anyone interested.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @jackkowalik, thank you for sharing this detailed discussion and your work with Ultralytics 🚀  
This is an automated response to help get things moving quickly, and an Ultralytics engineer will also assist you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Reply by glenn-jocher** [reply]
Thanks for sharing this, it’s a solid example of “downstream” improvement without touching the detector; one note for anyone trying to reproduce it in Ultralytics is that `Results.save_crop()` explicitly does **not** support OBB outputs, so for OBB you’ll want to crop from `result.orig_img` using `result.obb.xyxy` (axis-aligned enclosure) or `result.obb.xyxyxyxy` (rotated corners) as shown in the [Results API reference](https://docs.ultralytics.com/reference/engine/results/) and [Predict mode docs](https://docs.ultralytics.com/modes/predict/).

```python
from ultralytics import YOLO
import cv2

model = YOLO("yolo26n-obb.pt")
r = model("tile.jpg")[0]

if r.obb is not None:
    for box in r.obb.xyxy:  # (N, 4) axis-aligned boxes enclosing the rotated OBBs
        x1, y1, x2, y2 = box.int().tolist()
        crop = r.orig_img[y1:y2, x1:x2]  # numpy BGR
        crop_gray = cv2.cvtColor(crop, cv2.COLOR_BGR2GRAY)
```

Also, by default the post-NMS `Results` object exposes `result.obb.cls`/`result.obb.conf` (top-1 per detection); if you need a full per-class score vector per detection you’ll likely need to work from the raw model outputs before post-processing in a custom pipeline. If you want broad community testing/benchmarking, please keep it public in this discussion (we can’t coordinate private evaluation keys or off-thread support), and feel free to share a minimal runnable repo/notebook plus exact eval protocol so others can compare apples-to-apples; the [OBB task docs](https://docs.ultralytics.com/tasks/obb/) are a good common reference point.

**Reply by jackkowalik** [reply]
Thanks for the response and the cropping guidance! We're using `obb.xyxyxyxy` converted to axis-aligned enclosures, so same approach as your snippet.

Ran YOLO26l-OBB on the same DOTA v1.0 val set for comparison:

```
YOLOv8l-OBB  (50,348 detections, 5-fold scene-level CV)
                          Direct    Bolt-On
Weighted F1               0.9925    0.9929
Macro F1                  0.9826    0.9827

  helicopter              0.502  →  0.916   (+0.414)
  plane                   0.976  →  0.998   (+0.022)


YOLO26l-OBB  (49,414 detections, 5-fold scene-level CV)
                          Direct    Bolt-On
Weighted F1               0.9943    0.9947
Macro F1                  0.9891    0.9899

  baseball-diamond        0.994  →  0.997   (+0.003)
  ground-track-field      0.990  →  0.993   (+0.002)
  swimming-pool           0.998  →  1.000   (+0.002)
```

YOLO26 is a stronger baseline. It already handles helicopter at 0.966 where YOLOv8 was at 0.502. Encoder still reduces remaining error on both, no class degrades on either. Models with more real headroom for improvement achieve a higher reduction in error rate.

Reproducible script here: https://gist.github.com/jackkowalik/f354289a8892fe7d8d99e66da1b37eea

**Reply by glenn-jocher** [reply]
Good comparison—note you can use the cached axis-aligned enclosure directly via [`result.obb.xyxy`](https://docs.ultralytics.com/reference/engine/results/#property-ultralytics-engine-results-obb-xyxy) (no need to convert from `xyxyxyxy` unless you specifically want corner geometry), and it’s worth clipping to image bounds to avoid empty/invalid crops on edge cases:

```python
h, w = r.orig_img.shape[:2]
for b in r.obb.xyxy:
    x1, y1, x2, y2 = b.round().int().tolist()
    x1, y1, x2, y2 = max(0, x1), max(0, y1), min(w, x2), min(h, y2)
    if x2 > x1 and y2 > y1:
        crop = r.orig_img[y1:y2, x1:x2]
```

If you want others to reproduce quickly from your [gist script](https://gist.github.com/jackkowalik/f354289a8892fe7d8d99e66da1b37eea), please pin the exact `ultralytics` version and include `yolo checks` output, and (optionally) add standard [`task=obb` val metrics](https://docs.ultralytics.com/tasks/obb/) alongside the matched-detection CV so readers can separate detector gains from the downstream re-classifier.

**Reply by jackkowalik** [reply]
Updated the gist to pin ultralytics==8.4.21 with environment details. Switched cropping to obb.xyxy with bounds clipping per your suggestion. Standard OBB val for both models on DOTAv1 val:

yolo val model=yolov8l-obb.pt data=DOTAv1.yaml split=val device=0

                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)
                   all        458      28853      0.816      0.617      0.702      0.564
            helicopter         14         73      0.903      0.510      0.654      0.502


yolo val model=yolo26l-obb.pt data=DOTAv1.yaml split=val device=0

                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)
                   all        458      28853      0.845      0.621      0.718      0.578
            helicopter         14         73      0.831      0.534      0.661      0.465
Detection is identical between baseline and bolt-on in our benchmark. The same model, same weights, same conf threshold, and same cached detections are used. The "YOLO Direct" column is YOLO's own obb.cls on those matched detections. Any difference between "Direct" and "Bolt-On" is the result of the downstream reclassifier.

**Comment by jackkowalik** [comment]
Followed up on the full per-class score vector idea. Hooked into the raw model output (indices 4:19 of the pre-NMS tensor) and matched back to post-NMS detections by class score. Updated the gist with `--full-scores` flag.

YOLOv8l-OBB with full 15-dim class scores instead of one-hot:

```
                          Full-scores baseline    Full-scores + encoder
Weighted F1               0.9910                  0.9927
```

Encoder improves classification when the baseline has access to the complete class distribution. 18.4% error reduction over the full-scores baseline.

Tried the same with YOLO26l-OBB but the raw scores don't work the same way. YOLO26's one-to-one prediction head produces final detections directly without NMS, so the intermediate activations aren't interpretable class distributions. LightGBM on YOLO26 raw scores only reaches 0.9814 vs 0.9943 direct, confirming they're not comparable. Added a note about this in the gist docstring.

**Reply by glenn-jocher** [reply]
Thanks for the follow-up—your observation is consistent with the architectural shift: YOLOv8-OBB still has a post-process (NMS) stage where “pre-NMS class scores” can be treated as a usable class distribution, while YOLO26’s end-to-end, NMS-free head doesn’t expose an equivalent interpretable per-class probability vector for each final detection, so feeding those intermediate activations to LightGBM won’t be comparable; the [YOLO26 vs YOLOv8 comparison](https://docs.ultralytics.com/compare/yolo26-vs-yolov8/) has a bit more background on that design difference. If you’d like others to reproduce/try this publicly, please paste a minimal script + expected outputs directly here (we can’t do email/private-key coordination), and linking it as a small example PR against the [Ultralytics examples folder](https://github.com/ultralytics/ultralytics/tree/main/examples) would make it much easier for the community to test across datasets and [OBB](https://docs.ultralytics.com/tasks/obb/) models.

**Reply by jackkowalik** [reply]
PR submitted with a public evaluation key. No signup is needed: https://github.com/ultralytics/ultralytics/pull/23825

Script, README, and expected outputs are in examples/OBB-External-Feature-Reclassifier/

**Reply by glenn-jocher** [reply]
Thanks—saw [PR #23825](https://github.com/ultralytics/ultralytics/pull/23825) and we’ll review it; please remove/rotate the committed evaluation key (even if “public”) and load it from user-provided config/env at runtime with a placeholder in-repo, and if you can make the example gracefully no-op when the key is missing (with the README showing the exact expected output), it’ll be much easier/safer to maintain per our [contribution guidelines](https://docs.ultralytics.com/help/contributing/).
